### PR TITLE
Support Nomad Variables

### PR DESCRIPTION
**Pull request by angrycub**
This PR extends consul-template to support NomadVariables. It adds 4 additional functions:

- nomadVars
- nomadVarsSafe
- nomadVar
- nomadVarExists

There is a circular dependency issue with Nomad, so the tests won't pass until Nomad has released a version that contains secure variables. I'm not sure how exactly to manage this, but I need to get the PR up at least.

### CODE REVIEW

**Comment by eikenb** [comment]
> [..] so the tests won't pass until Nomad has released a version that contains secure variables.

Is it upstream yet? We can do a commit-id update for now if necessary.

Do you want these merged before the 0.29.2 release I'm planning for this week?

**Comment by tgross** [comment]
> Is it upstream yet? We can do a commit-id update for now if necessary.

Everything we need to support this in the API is on `main`. We're expecting to release Nomad 1.4.0-beta in the next few weeks.

> Do you want these merged before the 0.29.2 release I'm planning for this week?

Maybe we hold off on that? When I sent @angrycub off on this particular adventure 😀 I wasn't keeping in mind the beta. We can always pin to a commit-id on the Nomad side for the beta release and then if there's anything that comes up in the beta window where we really hate the API and need to change it, we're not causing regressions on a released version of `consul-template`.

So what do you (@eikenb and @angrycub) this about the following:
* For this PR we pin the go.mod to a commit-id on Nomad.
* `consul-template` 0.29.2 ships without this.
* We ship Nomad 1.4.0-beta in the next few weeks, pinning our CT dependency on the head of this PR's branch.
* Sometime during the beta window we merge this PR.
* Nomad 1.4.0 GA pins CT to a commit-id on CT's main at that point, or a hypothetical 0.29.x that lands afterwards.

That'll prevent Nomad from causing extra release schedule pressure for y'all, and give us a little breathing room around the Nomad 1.4.0 beta as well.

**Comment by eikenb** [comment]
Thanks @tgross. Sounds like a good plan.

File: dependency/nomad_secure_variables_get_test.go
Comment (tgross) [inline dependency/nomad_secure_variables_get_test.go:128]: Rather than `fmt.Printf`'ing in the test helper, we should probably return the error here and `t.Fatal` (or `require.NoError`, as we have testify in this repo); there's no reasonable behavior for the test after this point.

File: template/template_test.go
Comment (tgross) [inline template/template_test.go:2031]: Is it at all useful to have some tests exercising the namespace argument or some of the field access around metadata vs items?

File: docs/templating-language.md
Comment (tgross) [inline docs/templating-language.md:1970]: I suspect helper methods are not-too-unintuitive for go devs but for operators it might be useful to specifically document an example of their use here.

File: dependency/nomad_secure_variables_structs.go
Comment (tgross) [inline dependency/nomad_secure_variables_structs.go:81]: This ends up allocating two slices instead of one because of the call to Keys. Can we do `for k, val := range v` instead? (Same applies to `Tuples()` below.)

File: dependency/dependency_test.go
Comment (tgross) [inline dependency/dependency_test.go:436]: This function has no callers, so we're probably missing cleanup functions in the tests.

File: dependency/nomad_secure_variables_list_test.go
Comment (tgross) [inline dependency/nomad_secure_variables_list_test.go:175]: It doesn't look like any of the `Fetch` calls include a namespace, so I suspect we don't have good test coverage of the non-default namespace code paths.

File: dependency/nomad_secure_variables_structs.go
Comment (angrycub) [reply dependency/nomad_secure_variables_structs.go:81]: My main goal in using `Keys()` to create the other two outputs is that I wanted key-sorted order into the respective slices without repeating the sort code, so that they would range in the same order that go-template would order the map output.  Go template will do an implicit sort of keys so that they range consistently.

Comment (tgross) [reply dependency/nomad_secure_variables_structs.go:81]: Oh of course! That makes sense.

File: dependency/nomad_secure_variables_list_test.go
Comment (angrycub) [reply dependency/nomad_secure_variables_list_test.go:175]: Currently, the secure variables functions are only configured for the current namespace.  As you mentioned synchronously, there is a need for namespace support for service discovery as well.

File: template/template_test.go
Comment (angrycub) [reply template/template_test.go:2032]: Added tests to exercise namespace and region

File: dependency/nomad_secure_variables_list_test.go
Comment (angrycub) [reply dependency/nomad_secure_variables_list_test.go]: Updated to support namespaces and region

File: docs/templating-language.md
Comment (tgross) [inline docs/templating-language.md:1817]: It might be nice to link Nomad variables to https://www.nomadproject.io/docs/concepts/variables (and Nomad service registrations to https://www.nomadproject.io/docs/job-specification/service#provider )

File: go.mod
Comment (tgross) [inline go.mod:19]: Before we merge, let's update this again to the current beta version (at least).

**Comment by tgross** [comment]
I've opened https://github.com/hashicorp/nomad/issues/14762 with a reminder to ourselves to update Nomad to point to the new release of CT once this has been merged.

File: .github/workflows/ci.yml
Comment (eikenb) [inline .github/workflows/ci.yml:29]: Is this going to be removed before merging?

Comment (angrycub) [reply .github/workflows/ci.yml:29]: Yeah, it was the only way to actually have the tests pass in the interim.  Once there are real releases, we will pull this out.

**Comment by angrycub** [comment]
Updated the Nomad API dependency to 1.4.1's /api version.  Reverted the test shenanigans. Things look to be green and ready to rock @eikenb, @tgross

File: docs/templating-language.md
Comment (tgross) [inline docs/templating-language.md:1901]:
```suggestion
The `nanoTime` type contains Unix nanoseconds since the epoch. Its String method
```

**Review by tgross — APPROVED** [review]
👍

File: .go-version
Comment (eikenb) [inline .go-version:1]: I'm moving all my projects to use go.mod for version info. So this file will be ignored and confusing. Would you mind removing it?

**Review by eikenb — CHANGES_REQUESTED** [review]
Everything looks good except that one file. Thanks.

File: .go-version
Comment (angrycub) [reply .go-version:1]: Will consul-template not be switching to the common release tooling?  The CRT actions use the .go-version file to determine what go version to install.

Comment (eikenb) [reply .go-version:1]: Consul-template already uses CRT. CRT provides a generic example config for the github action workflows that needs to be tweaked and maintained for each project. It is not a fixed script. I didn't want/need the detailed control of .go-version but I did need it to use the appropriate compiler version for the code (as is specified in go.mod). Github actions supports doing that and it was only a [minor change to the build.yml](https://github.com/hashicorp/consul-template/commit/47c9feceb89932789301ae679ee89de61b7011a5).

**Review by eikenb — APPROVED** [review]

**Comment by eikenb** [comment]
@angrycub... could you rebase on main/HEAD to get the go.mod build changes? This should allow the tests to pass (or at least fix what's currently failing). Thanks.

**Pull request merged**
